clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableEventReaper.java

```diff
@@ -0,0 +1,154 @@
+ * making sure to reap them if they complete normally or pass their deadline. This is done so that we enforce an upper
+ * bound on the amount of time the event logic will execute.
+ */
+public class CompletableEventReaper<T extends CompletableEvent<?>> {
+
+    private final Logger log;
+
+    /**
+     * List of tracked events that we are candidates to expire or cancel when reviewed.
+     */
+    private final List<T> tracked;
+
+    public CompletableEventReaper(LogContext logContext) {
+        this.log = logContext.logger(CompletableEventReaper.class);
+        this.tracked = new ArrayList<>();
+    }
+
+    /**
+     * Adds a new {@link CompletableEvent event} to track for later completion/expiration.
+     *
+     * @param event Event to track
+     */
+    public void add(T event) {
+        tracked.add(Objects.requireNonNull(event, "Event to track must be non-null"));
+    }
+
+    /**
+     * This method "completes" any {@link CompletableEvent}s that have either expired or completed normally. So this
+     * is a two-step process:
+     *
+     * <ol>
+     *     <li>
+     *         For each tracked event which has exceeded its {@link CompletableEvent#deadlineMs() deadline}, an
+     *         instance of {@link TimeoutException} is created and passed to
+     *         {@link CompletableFuture#completeExceptionally(Throwable)}.
+     *     </li>
+     *     <li>
+     *         For each tracked event of which its {@link CompletableEvent#future() future} is already in the
+     *         {@link CompletableFuture#isDone() done} state, it will be removed from the list of tracked events.
+     *     </li>
+     * </ol>
+     *
+     * <p/>
+     *
+     * This method should be called at regular intervals, based upon the needs of the resource that owns the reaper.
+     *
+     * @param currentTimeMs <em>Current</em> time with which to compare against the
+     *                      <em>{@link CompletableEvent#deadlineMs() expiration time}</em>
+     */
+    public void reapExpiredAndCompleted(long currentTimeMs) {
+        log.trace("Reaping expired events");
+
+        Consumer<CompletableEvent<?>> timeoutEvent = e -> {
+            TimeoutException error = new TimeoutException(String.format("%s could not be completed within its timeout", e.getClass().getSimpleName()));
+            long pastDueMs = currentTimeMs - e.deadlineMs();
+            log.debug("Completing event {} exceptionally since it expired {} ms ago", e, pastDueMs);
+            CompletableFuture<?> f = e.future();
+            f.completeExceptionally(error);
+        };
+
+        // First, complete (exceptionally) any events that have passed their deadline AND aren't already complete.
+        tracked
+                .stream()
+                .filter(e -> !e.future().isDone())
+                .filter(e -> currentTimeMs > e.deadlineMs())
+                .forEach(timeoutEvent);
+
```

Comment:
nit:
IMO, this is more readable, but feel free to leave it if you do not share my taste.
```suggestion
        tracked.stream()
            .filter(e -> !e.future().isDone())
            .filter(e -> currentTimeMs > e.deadlineMs())
            .forEach(timeoutEvent);
```